### DISCUSSION TITLE
how can i build gitbutler on nix

### DISCUSSION THREAD

**Discussion by ymolists** [gitbutlerapp/gitbutler#4340] [Q&A] [answered]
Hi folks.

Thank you so much for this amazing tool.

I am trying to see if i can build this tool on nix. I noticed there is a flake file [here](https://github.com/gitbutlerapp/gitbutler/blob/master/flake.nix) but nothing about how to use it to either install the tool or build using it. Can anyone please help ?

Best regards

**Comment by Byron** [comment]
I am glad you like it :)!

Maybe @ndom91 can help here.

**Comment by ndom91** [comment]
Hey yeah so this nix flake file just contains a development shell config. So you can run `nix develop` in the repo directory and it'll install the environtment necessary to run a dev instance. So then you can launch it with `pnpm tauri dev`, for example. 

To run a production version of GitButler on NixOS, I recommend using the appimage releases

**Reply by ndom91** [reply]
That reminds me, someone did submit `0.12.7` to nixpkgs recently, check out: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/by-name/gi/gitbutler/package.nix#L169

It's available in the `unstable` channel, i.e. something like `unstable.gitbutler` if you're using flakes and have unstable setup as an input

**Reply by ymolists** [reply]
> Hey yeah so this nix flake file just contains a development shell config. So you can run `nix develop` in the repo directory and it'll install the environtment necessary to run a dev instance. So then you can launch it with `pnpm tauri dev`, for example.

Thank you for the prompt reply. Really appreciated.

I actually did just that. But the i get a complete blank screen when i run that. i also tried following the developer docs [here](https://github.com/gitbutlerapp/gitbutler/blob/master/DEVELOPMENT.md) and tried running it by doing `./target/release/git-butler-nightly`. Before i try running the released version can you tell me if i am doing anything wrong please ?

**Reply by Byron** [reply]
Getting a blank screen is very unusual, but I'd expect something [to be visible in the logs](https://docs.gitbutler.com/development/debugging#logs). I hope that helps.

**Reply by ndom91** [reply]
@ymolists are you seeing error messages related to DMABuf and Mesa versions in the logs? The `mesa` version used as a dependency of `webkitgtk` in this development shell must match the one used on your host system. i.e. if you're using `mesa` from `24.05` nixpkgs or `unstable` nixpkgs, etc. that's got to match. If thats the issue you're running into, you can replace the `webkitgtk` part in the `common` list (passed to `buildInputs`) with this:

```diff
common = with pkgs; [
  ...
- webkitgtk
+ (webkitgtk.overrideAttrs (old: {
+    buildInputs = [ unstablePkgs.mesa ] ++ old.buildInputs;
+ }))
  ...
];
```

Remember to swap the `unstablePkgs` prefix with whatever nixpkg instance has your particular version of `mesa`.

I'm investigating how to make this dynamic in the flake atm, but haven't found a way to dynamically define the `buildInputs` based on the values from the host system yet

**Reply by ymolists** [reply]
> @ymolists are you seeing error messages related to DMABuf and Mesa versions in the logs?  -→ ./target/release/git-butler-nightly 

Here[1] is what i am getting on the console. not sure if it is the same error or no.


```
[1]
** (git-butler-nightly:2111391): WARNING **: 21:02:00.609: webkit_settings_set_enable_offline_web_application_cache is deprecated and does nothing.
2024-07-14T01:02:00.812009Z  INFO crates/gitbutler-tauri/src/main.rs:99: starting app version=0.0.0 name=GitButler Nightly
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer
2024-07-14T01:02:01.550609Z  INFO get_user: crates/gitbutler-tauri/src/users.rs:10: close time.busy=15.8µs time.idle=20.9µs
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer
2024-07-14T01:02:01.849453Z  INFO list_projects: crates/gitbutler-tauri/src/projects.rs:41: close time.busy=21.6µs time.idle=25.0µs
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer
AcceleratedSurfaceDMABuf was unable to construct a complete framebuffer

```

**Reply by ndom91** [reply]
Yeah so that seems to be the same error. It looks like you're running the nightly, can you elaborate on how you're running gitbutler here? Is it the version from `nixpkgs`? A snippet from your config would be helpful

**Reply by ymolists** [reply]
I am following the docs [here](https://github.com/gitbutlerapp/gitbutler/blob/master/DEVELOPMENT.md) on how to build the code locally. i have not tried running the released binaries yet.

here is what i tried 

```
 -→ cat /etc/lsb-release 
DISTRIB_CODENAME=uakari
DISTRIB_DESCRIPTION="NixOS 24.05 (Uakari)"
DISTRIB_ID=nixos
DISTRIB_RELEASE="24.05"
LSB_VERSION="24.05 (Uakari)"

-→ git clean -fdx
-→ git reset --hard
-→ git l -1
9e0a0728e - (HEAD -> master, origin/master, origin/HEAD) Merge pull request #4408 from gitbutlerapp/upgrade-windows-crate (Tue Jul 16 11:09:33 2024, Kiril Videlov krlvi@users.noreply.github.com)

 -→ nix develop 
warning: input 'rust-overlay' has an override for a non-existent input 'flake-utils'

-→ pnpm i
 -→ cargo build
...
   Compiling tauri-plugin-single-instance v0.0.0 (https://github.com/tauri-apps/plugins-workspace?branch=v1#76b0f657)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 1m 02s
```

after it was done i ran this 
```
-→ LOG_LEVEL=debug pnpm tauri dev 
> root@ tauri /home/ymo/3rdp/github/gitbutler/gitbutler-git/gitbutler.devbr
> tauri "dev"

     Running BeforeDevCommand (`pnpm dev`)

> root@ dev /home/ymo/3rdp/github/gitbutler/gitbutler-git/gitbutler.devbr
> turbo watch --filter @gitbutler/app dev --no-update-notifier

• Packages in scope: @gitbutler/app
• Running dev in 1 packages
• Remote caching disabled
• Packages in scope: @gitbutler/app
• Running dev in 1 packages
• Remote caching disabled
  × discovery failed: bad grpc status code: The operation was cancelled

 ELIFECYCLE  Command failed with exit code 1.
       Error The "beforeDevCommand" terminated with a non-zero status code.
 ELIFECYCLE  Command failed with exit code 1.

```
when the above failed i tried building the nightly as per the docs above

```
-→ pnpm tauri build --features devtools --config crates/gitbutler-tauri/tauri.conf.nightly.json
...
...
  Compiling tauri-plugin-context-menu v0.7.1 (https://github.com/c2r0b/tauri-plugin-context-menu?branch=main#d8dcbcb9)
   Compiling tauri-plugin-window-state v0.1.1 (https://github.com/tauri-apps/plugins-workspace?branch=v1#76b0f657)
   Compiling tauri-plugin-store v0.0.0 (https://github.com/tauri-apps/plugins-workspace?branch=v1#76b0f657)
    Finished `release` profile [optimized + debuginfo] target(s) in 4m 03s
        Warn Signing, by default, is only supported on Windows hosts, but you can specify a custom signing command in `bundler > windows > sign_command`, for now, skipping signing the installer...
    Bundling git-butler-nightly_0.0.0_amd64.AppImage (/home/ymo/3rdp/github/gitbutler/gitbutler-git/gitbutler.devbr/target/release/bundle/appimage/git-butler-nightly_0.0.0_amd64.AppImage)
       Error failed to bundle project: error running appimage.sh

-→ ll ./target/release/git-butler-nightly 
-rwxr-xr-x 2 ymo users 236181568 Jul 16 11:27 target/release/git-butler-nightly
  gitbutler.devbr  master [!] 
 -→ ./target/release/git-butler-nightly

```

**Reply by ndom91** [reply, marked as answer]
@ymolists so actually that's more or less worked, _except_ for the appimage bundling part at the end. I can't get this to work on NixOS either, but if you check out the directory `/home/ymo/3rdp/github/gitbutler/gitbutler-git/gitbutler.devbr/target/release/bundle/appimage/`, you should have the unbundled appImage ocntents there. You can use the nix [appimage-run](https://search.nixos.org/packages?channel=24.05&show=appimage-run&from=0&size=50&sort=relevance&type=packages&query=appimage) package to execute the unbundled directory with `appimage-run -w .` (to run the `cwd`).

Anyway, long story short, that's how you can run that build. But in general I'd just recommend using the premade `gitbutler` nixpkgs package and/or the AppImage that we ship. 

The AppImage should work out of the box, but in case you're using the nixpkgs version and are running a version of `mesa` not from `nixpkgs-24.05`, then you'll have to override the version of `webkitgtk` used by the gitbutler derivation, as it depends on `mesa` and it has to match the version installed on your system.

See my changes to the `gitbutlerapp/gitbutler` repositories `flake.nix`. Basically it's just using `unstable` in the `common` list instead of the `24.05` `pkgs` (you can ignore the `webkitgtk_4_1` removal):

![image](https://github.com/user-attachments/assets/1617b795-c71e-4858-9586-fdf91ff43283)

**Reply by ymolists** [reply]
@ndom91 i want to say i really appreciate you taking the time to do this. to be honest i am a newb in nixos and i i did not running an app image in nix via appimage-run even existed :-) thank you so much for an amazing TIL moment ! i felt i could get it to work by just rebuilding it !

I now have git-butler running after i used the appimage-run on the released binary. for posterity this[1] is what i did to change that flake.nix file

Best regards

```
[1]
{
  inputs = {
    unstable.url = "nixpkgs/nixos-unstable";
    nixpkgs.url = "github:nixos/nixpkgs/nixos-24.05";
    flake-utils.url = "github:numtide/flake-utils";

    rust-overlay = {
      url = "github:oxalica/rust-overlay";

      inputs = {
        nixpkgs.follows = "nixpkgs";
        # flake-utils.follows = "flake-utils";
      };
    };
  };
  outputs = { self, nixpkgs, unstable, flake-utils, rust-overlay }:
    flake-utils.lib.eachDefaultSystem
      (system:
        let
          overlays = [ (import rust-overlay) ];
          pkgs = import nixpkgs {
            inherit system overlays;
          };

          unstablePkgs = import unstable {
            inherit system overlays;
          };

          rustToolchain = pkgs.pkgsBuildHost.rust-bin.fromRustupToolchainFile ./rust-toolchain.toml;

          common = with unstablePkgs; [
            appimage-run
            
            gtk3
            glib
            glib-networking
            dbus
            openssl_3
            librsvg
            gettext
            libiconv
            libsoup_3
            webkitgtk
            #webkitgtk_4_1
            nodejs_20
            corepack_20
          ];

          # runtime Deps
          libraries = with pkgs;[
            cairo
            pango
            harfbuzz
            gdk-pixbuf
          ] ++ common;

          # compile-time deps
          packages = with pkgs; [
            curl
            wget
            pkg-config
            rustToolchain
          ] ++ common;
        in
        with pkgs;
        {
          devShells.default = mkShell {
            nativeBuildInputs = packages;
            buildInputs = libraries;

            shellHook =
              ''
                export LD_LIBRARY_PATH=${pkgs.lib.makeLibraryPath libraries}:$LD_LIBRARY_PATH
                export XDG_DATA_DIRS=${pkgs.gsettings-desktop-schemas}/share/gsettings-schemas/${pkgs.gsettings-desktop-schemas.name}:${pkgs.gtk3}/share/gsettings-schemas/${pkgs.gtk3.name}:$XDG_DATA_DIRS
                export GIO_MODULE_DIR="${pkgs.glib-networking}/lib/gio/modules/"
              '';
          };
        }
      );
}
```
